src/main/kotlin/no/nav/syfo/personstatus/infrastructure/database/repository/PersonOversiktStatusRepository.kt

```diff
@@ -131,12 +132,88 @@ class PersonOversiktStatusRepository(private val database: DatabaseInterface) :
+            it.setString(1, personident.value)
+            it.executeQuery().toList { toPPersonOversiktStatus() }
+        }
+        return personoversiktStatus.firstOrNull()?.toPersonOversiktStatus()
+    }
+
+    override fun getOrCreatePersonOversiktStatusIfMissing(personident: PersonIdent): PersonOversiktStatus {
+        return database.connection.use { connection ->
+            connection.getPersonOversiktStatus(personident)
+                ?: PersonOversiktStatus(fnr = personident.value).also {
+                    connection.createPersonOversiktStatus(
+                        commit = true,
+                        personOversiktStatus = it,
+                    )
+                }
+        }
+    }
+
+    override fun lagreVeilederForBruker(
+        veilederBrukerKnytning: VeilederBrukerKnytning,
+        tildeltAv: String,
+    ) {
         database.connection.use { connection ->
-            val personoversiktStatus = connection.prepareStatement(GET_PERSON_OVERSIKT_STATUS).use {
-                it.setString(1, personident.value)
-                it.executeQuery().toList { toPPersonOversiktStatus() }
+            val existingVeilederAndEnhet = connection.getExistingVeilederAndEnhet(veilederBrukerKnytning)
+            if (existingVeilederAndEnhet == null) {
+                throw SQLException("lagreVeilederForBruker failed, no existing personoversiktStatus found.")
+            } else if (existingVeilederAndEnhet.veileder != veilederBrukerKnytning.veilederIdent) {
+                connection.updateVeileder(veilederBrukerKnytning, existingVeilederAndEnhet)
+                connection.addVeilederHistorikk(existingVeilederAndEnhet, veilederBrukerKnytning, tildeltAv)
+                connection.commit()
             }
-            return personoversiktStatus.firstOrNull()?.toPersonOversiktStatus()
+        }
+    }
+
+    private fun Connection.getExistingVeilederAndEnhet(veilederBrukerKnytning: VeilederBrukerKnytning) =
+        this.prepareStatement(GET_TILDELT_VEILEDER_QUERY).use {
+            it.setString(1, veilederBrukerKnytning.fnr)
+            it.executeQuery().toList {
+                VeilederAndEnhet(
+                    getInt("id"),
+                    getString("tildelt_veileder"),
+                    getString("tildelt_enhet")
+                )
+            }
+        }.firstOrNull()
+
+    private fun Connection.updateVeileder(
+        veilederBrukerKnytning: VeilederBrukerKnytning,
+        existingVeilederAndEnhet: VeilederAndEnhet,
+    ) {
+        val rowCount = this.prepareStatement(UPDATE_TILDELT_VEILEDER_QUERY).use {
+            it.setString(1, veilederBrukerKnytning.veilederIdent)
+            it.setObject(2, Timestamp.from(Instant.now()))
+            it.setInt(3, existingVeilederAndEnhet.id)
+            it.executeUpdate()
+        }
+        if (rowCount != 1) {
+            throw SQLException("lagreVeilederForBruker failed, expected single row to be updated.")
```

Comment:
Liten typo som henger igjen fra copy/paste😊
```suggestion
            throw SQLException("updateVeileder/updateTildeltVeileder failed, expected single row to be updated.")
```